lib/src/lib.rs

```diff
@@ -1,24 +1,232 @@
+    lock_file: Option<PathBuf>,
+}
+
+impl Niksi {
+    /// Constructs a new [`NiksiBuilder`].
+    /// See the builders documentation for more information.
+    pub fn builder() -> NiksiBuilder {
+        NiksiBuilder::new()
+    }
+
+    /// Outputs the JSON for .devcontainer.json as a String.
+    /// Fails if Serialization fails.
+    pub fn devcontainer_json(&self) -> Result<String, serde_json::Error> {
+        serde_json::to_string_pretty(&DevContainer::from(self.config.clone()))
+    }
+
+    /// Outputs the overrides.nix file with paths, name and version set according to the
+    /// [`NiksiConfig`] provided.
+    pub fn overrides_nix(&self) -> String {
+        format!(
+            r#"{{pkgs, ...}}: {{
+                paths = with pkgs; [{}];
+                name = "{}";
+                tag = "{}";
+            }}"#,
+            self.config.packages.join(" "),
+            &self.config.name,
+            &self.config.version,
+        )
+    }
+
+    /// Builds the docker image as specified in the [`NiksiConfig`] by:
+    /// 1. Downloading the template or a default template
+    /// 2. Writing the result of [`Niksi::overrides_nix`] to overrides.nix
+    /// 3. Copying the niksi.lock file over, if one exists
+    /// 4. Running `nix build`
+    /// 5. Returning the path of the build result (in the Nix store)
+    ///
+    /// Fails if any of the Nix commands or fs access fails
+    pub fn build(&self) -> Result<PathBuf> {
+        const DEFAULT_TEMPLATE: &str = "github:niksi-aalto/templates#plain";
+
+        let workdir = tempfile::tempdir().context("Failed to create tempdir to work in")?;
+
+        Command::new("nix")
+            .args([
+                "--extra-experimental-features",
+                "nix-command flakes",
+                "flake",
+                "init",
+                "-t",
+            ])
+            .arg(
+                self.config
+                    .template
+                    .clone()
+                    .map(|t| format!("github:niksi-aalto/templates#{}", t))
+                    .unwrap_or(DEFAULT_TEMPLATE.to_string()),
+            )
+            .current_dir(workdir.path())
+            .status()
+            .context("`nix flake init` failed with the specified template")?;
+
+        let overrides = self.overrides_nix();
+        let mut overrides_file = std::fs::File::create(workdir.path().join("overrides.nix"))
+            .context("Failed to create overrides.nix file")?;
+        overrides_file
+            .write_all(overrides.as_bytes())
+            .context("Failed to write overrides to overrides.nix")?;
+
+        if self.lock_file.is_file() {
+            std::fs::copy(self.lock_file.as_path(), workdir.path().join("flake.lock"))
+                .context("Failed to copy niksi lockfile")?;
+        }
+
+        let result = Command::new("nix")
+            .args([
+                "--extra-experimental-features",
+                "nix-command flakes",
+                "build",
```

Comment:
For the future: There should be an option for the user to turn on logging (`-L`) and to turn off remote builders (`--builders ""`).